packages/babel-traverse/test/replacement.js

```diff
@@ -112,6 +112,33 @@ describe("path/replacement", function () {
       });
       expect(visitCounter).toBe(1);
     });
+
+    // https://github.com/babel/babel/issues/12386
+    it("updates pathCache with the replaced node", () => {
+      const ast = parse(`() => (a?.b)?.c`, {
+        createParenthesizedExpressions: true,
+      });
+      traverse(ast, {
+        OptionalMemberExpression(path) {
+          path.node.type = "MemberExpression";
+          path.replaceWith(path.node.__clone());
+          path.parentPath.ensureBlock();
+
+          const aQuestionDotB = path.get("object").get("expression");
+          const aQuestionDotBNode = aQuestionDotB.node;
+          aQuestionDotBNode.type = "MemberExpression";
+          aQuestionDotB.replaceWith(aQuestionDotBNode.__clone());
+        },
+        ParenthesizedExpression(path) {
+          path.replaceWith(path.node.expression);
+        },
+      });
+      expect(generate(ast).code).toMatchInlineSnapshot(`
+        "() => {
+          return a.b.c;
```

Comment:
As expected `v7.12.7` returns
```js
() => {
          return (a.b).c;
};
```
here, which means the test captures the behaviour difference from the path cache.